File: 240929/뿌요뿌요/puyo-puyo.py
n = int(input())
area = []
for i in range(n):
    area.append(list(map(int, input().split())))
visited = []
for i in range(n):
    visited.append([0] * n)


def in_range(x, y):
    if 0 <= x < n and 0 <= y < n:
        return True
    return False

def can_go(x, y):
    if in_range(x, y) and area[x][y] == num and not visited[x][y]:
        return True
    return False

def dfs(x, y):
    global cnt
    dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]

    for dr in range(4):
        nx, ny = x + dx[dr], y + dy[dr]

        if can_go(nx, ny):
            cnt += 1
            visited[nx][ny] = 1
            dfs(nx, ny)


blocks = []
cnt = 0
for row in range(n):
    for col in range(n):
        num = area[row][col]
        if can_go(row, col):
            cnt = 1
            visited[row][col] = 1
            dfs(row, col)
        
            blocks.append(cnt)

cnt = 0
for item in blocks:
    if item >= 4:
        cnt += 1

print(cnt, max(blocks))

File: 250222/뿌요뿌요/puyo-puyo.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int n;
int area[100][100];
int cnt = 0;
int max_size = 1;
bool visited[100][100];
int tmp_cnt;
int dx[4] = { -1, 0, 1, 0 };
int dy[4] = { 0, 1, 0, -1 };

bool InRange(int x, int y) {
    return 0 <= x && x < n && 0 <= y && y < n;
}

bool CanGo(int x, int y, int num) {
    return InRange(x, y) && area[x][y] == num && !visited[x][y];
}

void DFS(int x, int y) {
    int num = area[x][y];

    for (int dr = 0; dr < 4; dr++) {
        int nx = x + dx[dr];
        int ny = y + dy[dr];

        if (CanGo(nx, ny, num)) {
            tmp_cnt++;
            visited[nx][ny] = true;
            max_size = max(max_size, tmp_cnt);
            DFS(nx, ny);

        }
    }
}

int main() {
    cin >> n;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            cin >> area[i][j];

    for (int i=0; i<n; i++)
        for (int j = 0; j < n; j++) {
            if (!visited[i][j]) {
                visited[i][j] = true;

                tmp_cnt = 1;
                DFS(i, j);

                if (tmp_cnt >= 4)
                    cnt++;
            }
        }

    cout << cnt << " " << max_size;
    
    return 0;
}